internals/overlord/checkstate/manager.go

```diff
@@ -48,42 +75,53 @@ func (m *CheckManager) NotifyCheckFailed(f FailureFunc) {
 
 // PlanChanged handles updates to the plan (server configuration),
 // stopping the previous checks and starting the new ones as required.
-func (m *CheckManager) PlanChanged(p *plan.Plan) {
-	m.mutex.Lock()
-	defer m.mutex.Unlock()
-
-	logger.Debugf("Configuring check manager (stopping %d, starting %d)",
-		len(m.checks), len(p.Checks))
-
-	// First stop existing checks.
-	for _, check := range m.checks {
-		check.cancel()
-	}
-	// Wait for all context cancellations to propagate and allow
-	// each goroutine to cleanly exit.
-	m.wg.Wait()
-
-	// Set the size of the next wait group
-	m.wg.Add(len(p.Checks))
-
-	// Then configure and start new checks.
-	checks := make(map[string]*checkData, len(p.Checks))
-	for name, config := range p.Checks {
-		ctx, cancel := context.WithCancel(context.Background())
-		check := &checkData{
-			config:  config,
-			checker: newChecker(config, p),
-			ctx:     ctx,
-			cancel:  cancel,
-			action:  m.callFailureHandlers,
+func (m *CheckManager) PlanChanged(newPlan *plan.Plan) {
+	// TODO: should figure out how to not stop/restart checks that haven't changed
+	//       so that unnecessary changes/tasks aren't created
+
+	// First cancel tasks of currently-running checks.
+	m.checksLock.Lock()
+	for name, data := range m.checks {
+		if data.cancel != nil {
+			data.cancel()
 		}
-		checks[name] = check
-		go func() {
-			defer m.wg.Done()
-			check.loop()
-		}()
+		delete(m.checks, name)
+	}
+	m.checksLock.Unlock()
+
+	// Update local reference to plan.
+	m.planLock.Lock()
+	m.plan = newPlan
+	m.planLock.Unlock()
+
+	// Start updated checks.
+	m.state.Lock()
+	defer m.state.Unlock()
+	for _, config := range newPlan.Checks {
+		m.performCheckChange(config)
+	}
+	if !m.ensureDone.Load() {
+		// Can't call EnsureBefore before Overlord.Loop is running (which will
+		// call m.Ensure for the first time).
+		return
 	}
```

Comment:
I need this as well.

Reply:
I wasn't totally happy with this, to be honest. I wonder if we should consider refactoring the higher layers a little so `PlanChanged` is only called after the overlord is running...